Add nodes and descriptions to tech tree's stencilNav node (fixes #138)

bthrall opened the pull request:
I primarily added nodes under the `JSContext Free Parsing` subtree for the Caching project, but also "On-Disk Baseline Code" and "Progressive Stencil" (suggestions from @moztcampbell ).

I'd like opinions on whether we are missing any ideas under the `JSContext Free Parsing` subtree, if the descriptions of the nodes isn't clear, or if we need to break the nodes up into smaller chunks (or group them into larger chunks).

bthrall committed:
Add nodes and descriptions to tech tree's stencilNav node (fixes #138)

bthrall commented:
I'd like to get reviews from @nbp, @arai-a, and @moztcampbell before merging.

moztcampbell commented on tech_tree/index.md:
```diff
@@ -117,6 +123,38 @@ functions][1681338]. Full support with our current design is challenging.
+
+## In-Memory Stencil Caching (stencil-nav) {#stencilNav}
+
+When caching JS bytecode through Necko, we can miss opportunities to reuse data due to how the cache policy for disk caching is set up. If we create a separate container for caching Stencils across navigations, we can potentially reduce bandwidth and load overhead when a subresource is reused. This also allows us to coalesce loading so that multiple requests for the same script will only need to  ...
```
"parse the data once". I suspect that in most cases necko will only load the data once, but we do end up with multiple parses in some cases, especially with navigation and tabs

nbp replied:
I think the core concept is here, however, it is not clear what "across navigation" means, nor what in "how the cache policy for disk caching" is problematic. Maybe take some inspiration form:

> When caching stencil to disk, we wait until the idle-time of the page to trigger the saving of its resources.
> If a user were to load the same resources before the idle time is triggered, we would not benefit from the caching which is in progress.
> Stencil for Navigation project aims at solving this issue by providing an in-memory cache which can be used ahead of saving the content to disk.

arai-a commented on tech_tree/index.md:
```diff
@@ -117,6 +123,38 @@ functions][1681338]. Full support with our current design is challenging.
 
+## On-Disk Baseline Code {#onDiskBaselineCode}
+
+Storing Baseline-compiled code on disk could speed up loading built-in code, like the Dev Tools.
```
maybe remove "loading" ?

the benefit here are the following, and both affects "execution" part:
  * perform baseline execution instead of interpreter execution from the beginning
  * skip baseline compilation happens between interpreter execution and baseline execution

also, let's use "DevTools" or "Developer Tools" to follow the document:
https://firefox-source-docs.mozilla.org/devtools-user/

nbp replied:
I think we should distinguish between saving baseline code for any page and saving baseline code for built-ins.
Both can be addressed concurrently if we wanted to.

bthrall replied:
@arai-a I'm not sure what this means:
>     * skip baseline compilation happens between interpreter execution and baseline execution
Do you mean that we can avoid doing the baseline compilation between interpreter execution and baseline execution because the compilation might be cached?
Or do you mean that the baseline compilation can happen asynchronously so that it is available when the transition from interpreter execution to baseline execution happens? (I think we might already do this?)

arai-a replied:
my understanding is the following.  let me know if I'm wrong:
  * the purpose here is to reuse the result of baseline compilation across page load or restart
  * by storing baseline-compiled code on disk:
    * the baseline compile can be skipped next time
    * the first execution can be performed with baseline, instead of interpreter
  * so, the effect is that, we can skip 2, 3, and 4 in the following
    1. load JavaScript source or XDR from disk or network
    2. compile/decode JavaScript source
    3. run in interpreter
    4. baseline compile
    5. execute baseline code
    6. ion compile
    7. execute ion code

and "loading built-in code" sounds like step 1.

moztcampbell commented on tech_tree/index.md:
```diff
@@ -117,6 +123,38 @@ functions][1681338]. Full support with our current design is challenging.
+
+The current caches are based around the Necko alternate-datastream API and some use-cases may be better served by an in memory cache similar to used for CSS and images. An in-memory cache could be used to store JIT code, CSS, and Images instead of having a separate cache for each type of data.
+
+We hope to make policies consistent across resources and gather the eviction and insertion policies into a single API.
```
We also hope to share code that decides when it is appropriate to even use cached data for a given request.

moztcampbell reviewed (approved):
Thanks for flushing this out a bit more. It is great to see the pieces layed out in more detail so that we can continue to understand how various ideas relate.

arai-a commented on tech_tree/index.md:
```diff
@@ -117,6 +123,38 @@ functions][1681338]. Full support with our current design is challenging.
+
+## Unified Subresource API {#unifiedSubresourceApi}
+
+The current caches are based around the Necko alternate-datastream API and some use-cases may be better served by an in memory cache similar to used for CSS and images. An in-memory cache could be used to store JIT code, CSS, and Images instead of having a separate cache for each type of data.
```
s/Images/images/ , or s/images/Images/

arai-a reviewed (approved):
Looks good :)

nbp commented on tech_tree/index.md:
```diff
@@ -33,6 +33,12 @@ index.
 
 For example, Frozen could be inlined into the tag, removing a dereference
 
+## Progressive Stencil {#progressiveStencil}
+
+Once we can share scopes alongside Stencils, the cost of instantiating a Stencil should be significantly reduced, and we can more easily only allocate BaseScripts as needed, instead of all of them at the start. This could lead to a performance benefit when we want a script quickly because we can load and parse more of it lazily.
+
+See [this bug](https://bugzilla.mozilla.org/show_bug.cgi?id=progressive-stencil).
```
Looking at the bug description, and the remaining open dependency.
I suggest changing this description to:

Make it possible to generate stencil progressively, by being either on-demand or concurrently, such that we can reduce the first-execution latency of scripts during page load.

moztcampbell replied:
Sorry @bthrall I think I mixed up the bugs. I was thinking about https://bugzilla.mozilla.org/show_bug.cgi?id=run-from-stencil when I was talking about that stuff. Nicolas is right that 'progressive-stencil' was actually the delazification work he did.

nbp commented on tech_tree/index.md:
```diff
@@ -117,6 +123,38 @@ functions][1681338]. Full support with our current design is challenging.
+
+By compressing the JIT code that is stored in the Necko on-disk cache, we might be able to reduce the bandwidth needed to save and restore the cache, and reduce size on disk.
+
+To get the necessary performance improvements of this feature, we will want to do the compression in a prioritized background thread.
```
I am not sure this last statement is necessary. This sounds like an implementation detail.

nbp commented on tech_tree/index.md:
```diff
@@ -117,6 +123,38 @@ functions][1681338]. Full support with our current design is challenging.
+
+To get the necessary performance improvements of this feature, we will want to do the compression in a prioritized background thread.
+
+## Stencil Navigation Scheduling {#stencilNavScheduling}
+
+Once we can parse without a JSContext, we can provide richer information for scheduling tasks via Firefox's Task Controller. We can convert the state machine for loading a script into tasks that are scheduled individually with proper dependencies and priorities, and we can use priority to affect which subresources are loaded first.
```
Well …, we already have a list to order how resources are loaded. As this is necessary for implementing `defer` and `async` `script` tags.

Before Stencil, and JSContext decoupling, if we wanted to preload any JS resources we would have to create a global (window) object. This object is quite heavy weight and has a lot of implied consequences, which make this approach impractical.

By removing the GC & JSContext from the parsing tasks, we enabled the DOM to guess what the user might visit next, and preload ahead of any user interactions the JS resources needed to speed-up the transition time to the next page, and not only the DOM of the pages.

moztcampbell replied:
With this scheduling discussion, I was hoping to touch on "loading a js resource" is actually a series of sub-tasks around fetching bytes, parsing, instantiation, etc. Those can all be expressed in TaskController and given some adaptiveness as needed.

bthrall replied:
Should we have a feature for preloading JS resources, since stencilNavScheduling is not going to cover that?

nbp commented on tech_tree/diagram.mmd:
```diff
@@ -74,15 +78,32 @@ flowchart LR
 
+    onDiskBaselineCode[/"On-Disk Baseline Code"/]
+    click onDiskBaselineCode "#onDiskBaselineCode"
+    stencilNav --> onDiskBaselineCode
+    inBinarySelfHostedBaseline --> onDiskBaselineCode
```
Why would this depend on `stencilNav` (in-memory cache)?
Why would this depends on self-hosted code?

We can do baseline code caching without having implemented it for self-hosted code.

moztcampbell replied:
These links came up because we weren't sure it was helpful to break down some items too small. For example, the in-binary self-hosted baseline was incorporating the decoupling of JSContext from the generated baseline code. We could try other structures of the graph, but it seems the practical path we take to getting on-disk baseline will have those other two implemented first

nbp commented on tech_tree/diagram.mmd:
```diff
@@ -74,15 +78,32 @@ flowchart LR
+
     root --> contextFreeParsing["JSContext Free Parsing"]
-    stencilNav --> contextFreeParsing
+    contextFreeParsing -.-> stencilNav
```
This edge is nice to have but not necessary.

nbp commented on tech_tree/diagram.mmd:
```diff
@@ -74,15 +78,32 @@ flowchart LR
+    click decoupledCaching "#decoupledCaching"
+    stencilNav --> stencilNavScheduling[/"Stencil Navigation Scheduling"/]
+    click stencilNavScheduling "#stencilNavScheduling"
+    stencilNavScheduling --> compressDiskCache[/"Compressed On-Disk Caching"/]
```
I do not think there is anything other project preventing us from using compression for on-disk storage today.

moztcampbell replied:
The reason we did not turn on compression today (code is landed behind a pref) is that it happens on main thread because both Necko and Parsing bounce between the main-thread. A practical form of compression requires it to run off thread which is a bit silly if we keep dispatching back and forth. We might be able to do it sooner once arai's work on using TaskController for parsing is done, but again that is being considered within the umbrella of stencilNavScheduling.

nbp commented on tech_tree/diagram.mmd:
```diff
@@ -74,15 +78,32 @@ flowchart LR
+    click stencilNavScheduling "#stencilNavScheduling"
+    stencilNavScheduling --> compressDiskCache[/"Compressed On-Disk Caching"/]
+    click compressDiskCache "#compressDiskCache"
+    stencilNav --> unifiedSubresourceApi["Unified Subresource API"]
+    click unifiedSubresourceApi "#unifiedSubresourceApi"
```
unifiedSubresourceApi is how we wanted to implement the stencilNav. I do not think splitting it in 2 makes sense.

moztcampbell replied:
The main focus of stencilNav is to decouple from ScriptLoader and specific realms. That probably makes sense regardless of when we get other sub resources in the same cache.

nbp commented on tech_tree/diagram.mmd:
```diff
@@ -74,15 +78,32 @@ flowchart LR
+
+    stencilNav[/"In-Memory Stencil Caching (stencil-nav)"/]
+    click stencilNav "#stencilNav"
+    stencilNav --> decoupledCaching[/"Decoupled Script Caching"/]
+    click decoupledCaching "#decoupledCaching"
+    stencilNav --> stencilNavScheduling[/"Stencil Navigation Scheduling"/]
```
I do not think `stencilNav` blocks `decoupledCaching` nor `stencilNavScheduling` (which maybe should be named `stencilPreloading`)
However, `decoupledCaching` might block `stencilNav` if we wanted to re-implement the Necko cache using the navigation cache. (which is not necessary)

mgaudet commented:
This is doubly-approved; time to land?

bthrall commented:
I'd like to address some of the comments before landing, but I won't get to it until next week, if that's okay.

bthrall committed:
Incorporate comments.

The links from stencilNav and inBinarySelfHostedBaseline to
onDiskBaselineCode are because the practical path to onDiskBaselineCode
will most likely have those implemented first, even if
onDiskBaselineCode doesn't strictly require them.

bthrall committed:
Sync up dashed-line links with manual styling.

Links that are supposed to be dashed-line in the MermaidJS are not
rendered that way in the 'elk' renderer, so we manually style them as
dashed-lines in CSS.

bthrall commented:
I've updated the diagram based on everyone's comments, and tried to disentangle the diagram.
